UART_driver/Core/Inc/CircularBuffer.hpp

```diff
@@ -0,0 +1,34 @@
+#ifndef CIRCULAR_BUFFER_HPP
+#define CIRCULAR_BUFFER_HPP
+
+#include <stdint.h>
+
+class CircularBuffer {
+    public:
+        // Constructor. `buf` is the buffer to be used as a circular buffer
+        CircularBuffer(uint8_t* buf, uint16_t size);
+
+        // Peek at a byte in the buffer
+        uint8_t peek(uint8_t& res, uint16_t dist);
+
+        // Read bytes from the buffer
+        uint8_t read(uint8_t* buf, uint16_t dist);
+
+        // Write byte to the buffer
+        uint8_t write(uint8_t byte);
+
+        // Returns true if the buffer has space
+        bool hasSpace();
+
+        // Returns the size of the current data available
+        uint16_t getDataSize();
```

Comment:
just by reading the function name and comment, I am not entirely sure I know what you mean... do you mean how many elements are currently in the circular buffer? Do you mean the size of the actual data type itself? Seems a little ambiguous, please reconsider the name